bee-message/src/address/mod.rs

```diff
@@ -67,4 +73,63 @@ impl Address {
         // SAFETY: encoding can't fail as `self` has already been validated and built.
         bech32::encode(hrp.as_ref(), self.pack_to_vec().to_base32(), Variant::Bech32).unwrap()
     }
+
+    ///
+    pub fn unlock(
+        &self,
+        unlock_block: &UnlockBlock,
+        inputs: &[(OutputId, &Output)],
+        context: &mut ValidationContext,
+    ) -> Result<(), ConflictReason> {
+        match (self, unlock_block) {
+            (Address::Ed25519(ed25519_address), UnlockBlock::Signature(unlock_block)) => {
+                if context.unlocked_addresses.contains(self) {
+                    return Err(ConflictReason::SemanticValidationFailed);
+                }
+
+                let Signature::Ed25519(signature) = unlock_block.signature();
+
+                if signature.is_valid(&context.essence_hash, ed25519_address).is_err() {
+                    return Err(ConflictReason::InvalidSignature);
+                }
+
+                context.unlocked_addresses.insert(*self);
+            }
+            (Address::Ed25519(_ed25519_address), UnlockBlock::Reference(_unlock_block)) => {
+                // TODO actually check that it was unlocked by the same signature.
```

Comment:
Will this be a separate PR?

Reply:
Same https://github.com/iotaledger/bee/issues/1272